Update :server connect form to pick database

OskarDamkjaer opened the pull request:
Introduces a `db` url param for selecting a database and `dbms` as an alternative param to connectURL.

Prefills form from the url like so:
![bild](https://user-images.githubusercontent.com/10564538/93226834-0b88c080-f774-11ea-8ef2-8cd4260e07b2.png)

OskarDamkjaer commented on src/browser/components/buttons/index.jsx:
```diff
@@ -217,27 +217,27 @@ export const FormButton = props => {
   if (icon && label) {
     return (
-      <ButtonType {...rest} type="button">
+      <ButtonType type="button" {...rest}>
```
The other order overrides type=submit

OskarDamkjaer commented on src/shared/modules/connections/connectionsDuck.js:
```diff
@@ -37,7 +39,6 @@ import {
 import { inWebEnv, USER_CLEAR, APP_START } from 'shared/modules/app/appDuck'
 
 export const NAME = 'connections'
-export const ADD = 'connections/ADD'
```
Unused

OskarDamkjaer commented on src/shared/modules/connections/connectionsDuck.js:
```diff
@@ -165,25 +163,17 @@ const removeConnectionHelper = (state, connectionId) => {
+    allConnectionIds: allConnectionIds.includes(id)
+      ? allConnectionIds
+      : [...allConnectionIds, id]
   }
```
These method cleanup is think are cleaner, and in passing they also fixed a small bug that left a connection ID "undefined" in the connectionsById state

OskarDamkjaer commented on src/shared/modules/discovery/discoveryDuck.js:
```diff
@@ -79,17 +79,21 @@ export const getBoltHost = state => {
+    { host: action.forceURL }
+  )
+
   if (typeof action.encrypted !== 'undefined') {
```
Is there a reason this was checked to be undefined rather than truthyness? @oskarhane

OskarDamkjaer commented on src/shared/modules/discovery/discoveryDuck.js:
```diff
@@ -112,9 +116,15 @@ export const discoveryOnStartupEpic = (some$, store) => {
+
       if (!passedURL || !passedURL.length) return action
       action.forceURL = decodeURIComponent(passedURL[0])
+      action.connectTo = passedDb && passedDb[0]
```
Right now db param is only respected if there's also a dbms/connectURL param. Just a note on behaviour

OskarDamkjaer replied:
The reason I didn't use useDb is because it think the same state already has a useDb. But connectTo is a bad name, so perhaps requestedUseDb is a better name?

OskarDamkjaer commented on src/shared/modules/discovery/discoveryDuck.js:
```diff
@@ -124,14 +134,12 @@ export const discoveryOnStartupEpic = (some$, store) => {
           {
             ...action,
             username,
-            password,
```
No longer keep password from url param

OskarDamkjaer commented on src/shared/modules/discovery/discoveryDuck.js:
```diff
@@ -124,14 +134,12 @@ export const discoveryOnStartupEpic = (some$, store) => {
             ...action,
             username,
-            password,
+            showDbField: !!action.connectTo,
```
Not sure if this boolean is the best way to signal if the field should show up or not, was thinking of just checking for the presence of connectTo, but that is a bit awkward when using the version check bellow since there's no value to prefill.

OskarDamkjaer replied:
Yeah that's a better name!

OskarDamkjaer commented on src/shared/modules/discovery/discoveryDuck.js:
```diff
@@ -156,7 +167,11 @@ export const discoveryOnStartupEpic = (some$, store) => {
             )
-            store.dispatch(updateDiscoveryConnection({ host })) // Update discovery host in redux
+            const supportsMultiDb =
+              parseInt((result.neo4j_version || '0').charAt(0)) >= 4
```
I'm thinking there's a better way to get version number? Or is this fine?

OskarDamkjaer replied:
So do you suggest we check that the version starts with 4 and is not running on aura?

OskarDamkjaer replied:
Also, good catch!

oskarhane replied:
Some aura editions do support multi-db, so it's tricky. I suggest you talk to Aura and ask which editions that has multi-db support and what version string those editions expose in the discovery endpoint.

OskarDamkjaer replied:
Asked just now!

OskarDamkjaer replied:
It seems aura only supports the system + 1 other database at this moment. I think it'd make sense to disable the field for aura, WDYT?

OskarDamkjaer commented on src/browser/modules/Stream/Auth/ConnectForm.tsx:
```diff
@@ -0,0 +1,235 @@
+
+  return (
+    <StyledConnectionForm onSubmit={onConnectClick}>
+      <FlexSpaceBetween>
```
Not a 100% happy with this styling, it feels more convoluted than it needs to be. I opted to make theses simple Flex components rather than inline style even if they are very simple, for consistency's sake

OskarDamkjaer commented on src/shared/modules/dbMeta/dbMetaDuck.js:
```diff
@@ -431,11 +433,26 @@ export const dbMetaEpic = (some$, store) =>
+                      store.dispatch(useDb(wantedDb.name))
+                    } else {
+                      // this will show the db not found frame
+                      store.dispatch(executeSingleCommand(`:use ${connectTo}`))
```
I thought this was pretty nice error handling of databases that weren't found. Thoughts?

oskarhane commented on src/browser/modules/Stream/Auth/ConnectForm.tsx:
```diff
@@ -0,0 +1,235 @@
+                data-testid="database"
+                onChange={props.onDatabaseChange}
+                value={props.database}
+                placeholder="default database"
```
Maybe change this to "Database to use" or something

OskarDamkjaer replied:
You mean the label? I think the current placeholder makes sense

oskarhane replied:
I don't :)
It's not a default db the users enters, it's the target db that the user wants to use when a connection is made.

OskarDamkjaer replied:
Yeah and that's why the place holder is there to show what happens if you leave the field empty! Right now it think it makes sense to have a field titled "Database to use" and the default/placeholder being "default database". Having both be "database to use" I think would lead to me thinking the field was required

oskarhane replied:
I see your reasoning, but placeholders are hints on what the users should enter, not to indicate value if left blank.
The label could suggest "leave empty for default)" or something to let the users know what happens if it's left empty.

oskarhane reviewed (commented):
Cool, well done!

I have a few comments on naming and UI.

### Connect frame UI
I don't think this looks particularly good to be honest. The form looks very unusual.
<img width="981" alt="Screen Shot 2020-09-18 at 15 05 12" src="https://user-images.githubusercontent.com/570998/93601342-59463880-f9c1-11ea-8f5c-e8001c8f64a2.png">

### Auto-connect on refresh
It looks like the auto-connect doesn't work anymore. Successfully connect. Refresh page. You should auto-connect.

### Continuously trying to `:use`
In the scenario of connecting successfully but using a db that doesn't exist for some reason, it looks like it's running `:use` on every dbmeta-tick.
<img width="1407" alt="Screen Shot 2020-09-18 at 15 14 24" src="https://user-images.githubusercontent.com/570998/93601775-f99c5d00-f9c1-11ea-9ca1-47a60984f261.png">

OskarDamkjaer commented:
Good catches, didn't see the screen shots you sent! Thanks a lot! :))

OskarDamkjaer commented:
Updated the form styling when I was thinking we'd always show the db field. Then it got really tall, but now when I know that's not the case I reverted to the older styling because it was simpler code wise and familiar.

OskarDamkjaer commented:
Not 100% sure the aura check works. What would be the best way to try it out? Not sure where to find the discovery response object

oskarhane commented on e2e_tests/integration/connect-form.spec.ts:
```diff
@@ -76,10 +76,15 @@ describe('Connect form', () => {
   })
 
-  it('can connect with bolt:// protocol', () => {
+  it.only('can connect with bolt:// protocol', () => {
```
Was this intentional?

OskarDamkjaer replied:
No, I missed it after debugging the test. Should have read over the files one more time, good catch!

oskarhane commented on src/shared/modules/discovery/discoveryDuck.js:
```diff
@@ -156,7 +168,11 @@ export const discoveryOnStartupEpic = (some$, store) => {
+
+            const isAura = getEnv(store.getState()) === CLOUD
+            const supportsMultiDb =
+              !isAura && parseInt((result.neo4j_version || '0').charAt(0)) >= 4
```
Aren't there Aura editions that support multi-db?

OskarDamkjaer replied:
From asking there's only editions with proper support for multi-db. At most they support the system-db in addition to the default one. So I think if it's important there, you'll have to use the url param

oskarhane commented on e2e_tests/integration/connect-form.spec.ts:
```diff
@@ -80,6 +80,11 @@ describe('Connect form', () => {
     cy.executeCommand(':clear')
     const boltUrl = 'bolt://' + stripScheme(Cypress.config('boltUrl'))
     cy.connect('neo4j', Cypress.config('password'), boltUrl)
+    cy.executeCommand('RETURN "Hello World";')
+    cy.contains('Hello World')
+    cy.reload()
+    cy.executeCommand('RETURN "Hello again";')
+    cy.contains('Hello again')
```
👍

oskarhane reviewed (commented):
Good changes!

Still has a few more issues though:

1. When choosing something other than the default db, there's a delay until next dbMeta to update the meta items in the sidebar. See gif.
![CleanShot 2020-10-01 at 08 46 30](https://user-images.githubusercontent.com/570998/94776833-b929df80-03c2-11eb-908c-52166783159a.gif)


1. Same thing when entering a db that doesn't exist for the user. It takes until next dbmeta tick to update the UI with the correct info. 
![CleanShot 2020-10-01 at 08 51 25](https://user-images.githubusercontent.com/570998/94777266-77e5ff80-03c3-11eb-8f1d-40475325bbce.gif)

Also, would it be possible to add e2e test for checking so browser actually switches to use the entered db?
And maybe the error case as well?

oskarhane commented on e2e_tests/integration/connect-form.spec.ts:
```diff
@@ -105,5 +111,68 @@ describe('Connect form', () => {
     })
+
+    if (isEnterpriseEdition()) {
+      it.only('shows correct metadata when using db field', () => {
```
No `.only` please

OskarDamkjaer replied:
I'll set up an on commit hook for this so I won't forget it again

oskarhane commented on src/shared/modules/dbMeta/dbMetaDuck.js:
```diff
@@ -339,132 +483,20 @@ export const dbMetaEpic = (some$, store) =>
-                  return Rx.Observable.of(null)
-                })
+          .mergeMap(() => {
+            return Rx.Observable.forkJoin([
+              getLabelsAndTypes(store),
+              clusterRole(store),
+              databaseList(store)
```
Even if one of these errors we'll need to keep the stream going, so I'f like to see some error catching here

oskarhane replied:
And if I remember correctly, if one fail the other once are cancelled as well, so I think you need to catch on each one of the inner ones.

oskarhane reviewed (commented):
Seems to behave as expected now, well done!
You just need to be extra cautious to make sure one failure in the `forkJoin` doesn't cancel the other ones + cancels the whole listener.

But other than that, 👍 

Ping me to review when you're catching any errors and I'll approve it

oskarhane reviewed (approved):
🌮 Merge on green tests 🌮
